### PR TITLE
Fix spatialite handling of JSON arrays

### PR DESCRIPTION
**Pull request by elpaso**
Fixes #21986

plus:

- fix multiple string keys with commas in value relation widget
- more robust JSON and array (un)marshalling
- uniform array representation in value relation widgets
- lot of test coverage
- automatic QVariant type conversions in JSON utils

### CODE REVIEW

File: python/core/auto_generated/fieldformatter/qgsvaluerelationfieldformatter.sip.in [hunk 1]
```diff
@@ -59,7 +59,7 @@ Constructor for QgsValueRelationFieldFormatter.
 
     static QStringList valueToStringList( const QVariant &value );
 %Docstring
-Utility to convert an array or a string representation of an array ``value`` to a string list
+Utility to convert an array or a string representation of an (C style: {1,2...}) array in the form ``value`` to a string list
```
Comment (m-kuhn) [inline python/core/auto_generated/fieldformatter/qgsvaluerelationfieldformatter.sip.in]: Should that be hstore instead of C style? (and list instead of array since it's either `QVariantList` or python `list`)?

No big issue, just wondering.

File: src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp [hunk 2]
```diff
@@ -164,11 +168,49 @@ QgsValueRelationFieldFormatter::ValueRelationCache QgsValueRelationFieldFormatte
 
 QStringList QgsValueRelationFieldFormatter::valueToStringList( const QVariant &value )
 {
+  // Note: the recommended way to pass a value is QVariant::StringList but
+  //       for back compatibility a string representation either in the form
```
Comment (m-kuhn) [inline src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp]: for backwards compatibility and also awesome formats like shp

**Comment by luipir** [comment]
@elpaso am I wron or this PR is not only related to spatialite and more with widget wrapper and json parsing?

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp [hunk 3]
```diff
@@ -70,21 +75,26 @@ QVariant QgsValueRelationWidgetWrapper::value() const
...
-    if ( layer()->fields().at( fieldIdx() ).type() == QVariant::Map )
+    QVariantList vl;
+    //store as QVariantList because it's json
+    for ( const QString &s : qgis::as_const( selection ) )
     {
-      QVariantList vl;
-      //store as QVariantList because it's json
-      for ( const QString &s : qgis::as_const( selection ) )
+      // Convert to proper type
+      const auto type { fkType() };
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: Might be a personal opinion, I prefer explicit over implicit because that improves code readability (i.e. I only use auto when it's very clear what it's about and the real type is crappy to write like e.g. iterators).

**Comment by elpaso** [comment]
> @elpaso am I wron or this PR is not only related to spatialite and more with widget wrapper and json parsing?

you are not wrong, but git commit messages must be short, I think I added the details.

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h [hunk 4]
```diff
@@ -115,6 +115,9 @@ class GUI_EXPORT QgsValueRelationWidgetWrapper : public QgsEditorWidgetWrapper
      */
     int columnCount() const;
 
+    //! Returns the variant type of the fk
```
Comment (m-kuhn) [inline src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h]: \since?

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp (hunk 3 above)
Comment (elpaso) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: Well, the good thing with `auto` is that if for some reason we change the return type in the future it *may* not be necessary to change other parts of the code.
I see your point, but since QtCreator show a nice tooltip with the actual type I think it's not big issue.

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h (hunk 4 above)
Comment (elpaso) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h]: it's private ... do we need since?

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp (hunk 3 above)
Comment (m-kuhn) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: The point is, github doesn't ;)

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h (hunk 4 above)
Comment (m-kuhn) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.h]: oops

File: src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp (hunk 2 above)
Comment (elpaso) [reply src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp]: good point, should we add tests with shapefiles?

Comment (m-kuhn) [reply src/core/fieldformatter/qgsvaluerelationfieldformatter.cpp]: there would certainly be nothing wrong about adding tests. Main point is: I don't want this code to look like "deprecated" because of a comment.

**Comment by m-kuhn** [comment]
@signedav what do you think about this?

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp (hunk 3 above)
Comment (nyalldawson) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: Also, I may be alone in this, but to me the ability to change return type without compilation errors is a huge step backwards. There's no telling what subtle bugs may arise as a result of doing this.

Comment (elpaso) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: Ok guys you have almost convinced me, I will use less auto.
 Pity because l was starting to like C++ ;)

Comment (nyalldawson) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: Just to further explain my concerns:

Let's say I have the c++ method

    QgsGeometry getValue() const;

It's well tested under python, with lots of unit tests covering all possible return values. In future someone comes along and decides that this method needs to be more flexible (and e.g. sometimes return QgsRectangles), so they change the signature:

    QVariant getValue() const;

All the existing python tests are fine, because sip auto-converts the QVariant return values to QgsGeometry. So it looks good, nothing has broken.

But buried somewhere in c++ is this usage:

   auto val = foo.getValue();
   if ( !val.isNull() )
   {
      ....
    }

Previously, this was testing QgsGeometry::isNull(), and worked as intended. By changing the return type to QVariant, this block is now testing QVariant::isNull(), which has a very different meaning! Bam, silently, very difficult to diagnose regression, despite being soaked by Python unit tests.

That's why auto is a step back -- if we used QgsGeometry = foo.getValue(), compilation would break and we'd see that that block needs fixing manually.

Comment (elpaso) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: Thanks for the explanation.
I think we need a 'soft' policy.
Would everyone agrees on this answer?
https://stackoverflow.com/a/6435036?stw=2

**Comment by signedav** [comment]
Checked the code @m-kuhn and it looks like a reasonable improvement to the JSON type implementation. I haven't found any issues. Thanks @elpaso

**Comment by elpaso** [comment]
@signedav thank you for reviewing (the logic of) my PR!

File: src/core/qgsjsonutils.cpp [hunk 5]
```diff
@@ -315,28 +315,69 @@ QString QgsJsonUtils::exportAttributes( const QgsFeature &feature, QgsVectorLaye
...
+        // Note: it's a double and not a float on purpose
+        v = item.get<double>();
+      }
+      else if ( item.is_string() )
+      {
+        v = QString::fromStdString( item.get<std::string>() );
+      }
+      else if ( item.is_null() )
+      {
+        // do nothing: the default is null
```
Comment (m-kuhn) [inline src/core/qgsjsonutils.cpp]: Maybe this would be better to have NULL instead of invalid?
```
v = QVariant( QVariant::Int );
```

Comment (elpaso) [reply src/core/qgsjsonutils.cpp]: @m-kuhn 
Like this?
```
  // Type not specified
  QCOMPARE( QgsJsonUtils::parseArray( R"([null, null])" ), QVariantList() << QVariant() << QVariant() );
  // Convert to int
  QCOMPARE( QgsJsonUtils::parseArray( R"([null, null])", QVariant::Type::Int ), QVariantList() << QVariant(QVariant::Type::Int) << QVariant(QVariant::Type::Int) );
```

Comment (m-kuhn) [reply src/core/qgsjsonutils.cpp]: That's what I was thinking:

```
for ( const QVariant &value : values )
{
  QVERIFY( value.isNull() );
  QVERIFY( value.isValid() );
}
```

Comment (elpaso) [reply src/core/qgsjsonutils.cpp]: @m-kuhn that won't work for `QgsJsonUtils::parseArray( R"([null, null])" )` because we if we do not know the expected type we cannot set it and `QVERIFY( QVariant().isValid() );` will fail.

Comment (elpaso) [reply src/core/qgsjsonutils.cpp]: ... but it's fine in the other case though.

Comment (m-kuhn) [reply src/core/qgsjsonutils.cpp]: Yes, that's also the case when e.g. converting `qgis.core.NULL` to a `QVariant`, we just use `QVariant::Int` as fallback type (meaning possibly wrong type information under these circumstances).

Comment (elpaso) [reply src/core/qgsjsonutils.cpp]: @m-kuhn please have a look to my last commit, if that interprets your thoughts.

File: tests/src/core/testqgsjsonutils.cpp [hunk 6]
```diff
@@ -91,19 +91,17 @@ void TestQgsJsonUtils::testJsonArray()
...
   // Nulls
-  QCOMPARE( QgsJsonUtils::parseArray( R"([null, null])" ), QVariantList() << QVariant() << QVariant() );
-  QVERIFY( QVariant().isValid() );
   for ( const QVariant &value : QgsJsonUtils::parseArray( R"([null, null])" ) )
   {
     QVERIFY( value.isNull() );
-    // This would fail because the expected type is not set and an invalid QVariant is created:
-    // QVERIFY( value.isValid() );
+    QVERIFY( value.isValid() );
+    QCOMPARE( value, QVariant( QVariant::Type::Int ) );
```
Comment (m-kuhn) [inline tests/src/core/testqgsjsonutils.cpp]: Nice!

Now really out of curiosity:
I normally don't test for API behavior which is more a "side-effect" that could potentially change any time (let's assume it's replaced by `QVariant::Type::Dynamic` in the future). How do you feel about this kind of tests?

File: src/core/qgsjsonutils.cpp (hunk 5 above)
Comment (m-kuhn) [reply src/core/qgsjsonutils.cpp]: Yep, looking great. Just left a question that popped up, but good like this :+1:

File: tests/src/core/testqgsjsonutils.cpp (hunk 6 above)
Comment (elpaso) [reply tests/src/core/testqgsjsonutils.cpp]: Sorry, I'm not sure I understand :) 
Do you mean I should not test this case or that there is a better (more resilient) way to test it?

Comment (m-kuhn) [reply tests/src/core/testqgsjsonutils.cpp]: The first one. But I didn't mean "you shouldn't" and was more out for an answer to the question "what advantage do we expect from this test".

Comment (elpaso) [reply tests/src/core/testqgsjsonutils.cpp]: ok, maybe an obvious answer but the test is IMO important to test that the method behaves as expected. If I got your question right, we might better define (in `qgis.h` maybe) a `QVariant::Type::Something`  that we use as a fallback in case we want something to represent a  `NULL` but not invalid (i.e. undefined), and the use that alias in the tests (and in the code of course). 
Or perhaps go the `QgsVariant` way to handle `NULL` explicitely.

Comment (m-kuhn) [reply tests/src/core/testqgsjsonutils.cpp]: Then the question is, is returning an int "expected" or just "a side-effect". I think that no code consuming this API should rely on this fact (it should just test for `isValid()`, `isNull()` or rely on other sources of type information but not the NULL value itself. Hence the question: what's the advantage of testing that workarounds are kept the way they are. I hope I could explain a bit better :)

In short: Implementation-wise I would not change anything (I don't see a better approach), but I wonder if we should add tests to ensure leftovers of workarounds are "stable".

Comment (elpaso) [reply tests/src/core/testqgsjsonutils.cpp]: Ok, got it finally  :)
I agree that it should probably not be tested, isNull and isValid should be sufficient, I'll try to keep that in mind next time.
I find myself writing tests with two different goals, one is to check that the method works as expected (the Int here) and the other one is more API oriented and it is that it won't break in the future (kind of policy/insurance), I think that I should remove or comment the tests o in the first category, but it's not always easy to tell them apart.

Comment (m-kuhn) [reply tests/src/core/testqgsjsonutils.cpp]: Sounds good. I was just thinking loud here.

File: src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp (hunk 3 above)
Comment (m-kuhn) [reply src/gui/editorwidgets/qgsvaluerelationwidgetwrapper.cpp]: +1 for the soft policy.

> I think that one should use the auto keyword whenever it's hard to say how to write the type at first sight, but the type of the right hand side of an expression is obvious.

Just one minor correction, where I think there is a mismatch between examples and the above specification.

```cpp
std::shared_ptr<Foo> foo = std::make_shared<Foo>(); // while obvious, it's also easy to write at first sight
```
